### DISCUSSION TITLE
Is there an example of how to create a complete CRUD function using a custom repositories?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#7796] [Q&A]
I have a custom setup where I use the underlining YesSql connection to connect to the database using EntityCoreFramework. I have a table called "Locations" and I want to add a simple CRUD operation to that table from the admin panel.

I added the menu item to the admin menu and able to interact with the database. However, I am trying to figure out how to create a CRUD functions. I want to be able to add, edit, and create paged list for my locations.

Is there an example somewhere that demonstrate how to use the `DisplayManagement` module to create CRUD function? I am guessing I'll need to create a `LocationDisplayDriver` of some sort, but not sure what methods I would use to render an empty CreateViewModel or how to capture the POST request.

**Comment by hishamco** [comment]
If I'm not wrong @OrchardSkills refers to CRUD repo long time back, please have a look to https://github.com/OrchardSkills/OrchardSkills.OrchardCore.CRUD

**Reply by MikeAlhayek** [reply]
@hishamco is there a way to handle all this using a Display Driver instead? Or, is this the best way to handle content that are managed outside of YesSQL?

**Reply by hishamco** [reply]
I'm don't have deep knowledge about display drivers, but I can ping @deanmarcussen to answer this

**Reply by jsalgado10** [reply]
@CrestApps I am working on something similar but instead of creating a new table to store the records, I created a new content type and create a new controller to handle the CRUD operations. I use the AdminController in the OrchardCore.Contents module as a reference.

**Reply by MikeAlhayek** [reply]
@hishamco @deanmarcussen @jsalgado10  I tried to follow the example https://github.com/OrchardSkills/OrchardSkills.OrchardCore.CRUD to create my own CRUD. My controller can be found [here  ](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Controllers/LocationController.cs). I got everything working, except for the Edit page. I some how need to access the `Id` property so I can pass it the action attribute of the form. Currently, the [Model.Id](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Views/Locations/Edit.cshtml#L10) is always null and I am expecting the `Id` of the Edit model. Any idea on how can I get/set the model ID?

In the driver I am setting the ID using [AutoMapper ](https://automapper.org/) as you can see [here ](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Drivers/LocationDisplayDriver.cs)

Also, is there a way to simplify my code? I am hoping that the Edit filed can be reusable so I don't have to create a view for everything. Is this possible?

**Reply by jsalgado10** [reply]
@CrestApps from what I am seeing, you are passing a new instance of a Location class when building the editor on the Create HTTP Get method. However, there is nothing in the Location class that generates this id. 

If you look at the AdminController on the OrchardCore.Contents module, you could see how they are generating a new content record

`        public async Task<IActionResult> Create(string id)
        {
            if (String.IsNullOrWhiteSpace(id))
            {
                return NotFound();
            }

            var contentItem = await _contentManager.NewAsync(id);

            // Set the current user as the owner to check for ownership permissions on creation
            contentItem.Owner = User.Identity.Name;

            if (!await _authorizationService.AuthorizeAsync(User, Permissions.EditContent, contentItem))
            {
                return Forbid();
            }

            var model = await _contentItemDisplayManager.BuildEditorAsync(contentItem, _updateModelAccessor.ModelUpdater, true);

            return View(model);
        }`

Now if you are wanting to use you own IUnitOfWork, then you could create a dummy record first then pass that when building the editor

var dummyLocation = create new record on Db using IUnitOfWork
IShape shape = await _displayManager.BuildEditorAsync(dummyLocation, this, true);

**Reply by jsalgado10** [reply]
Any reason you need the Id property when creating a new record? I would think that id will be auto generated when it is added to the database. I chose to create a new content type with the require properties on OrchardCore CMS, then use those content type definitions when creating the record. This keeps everything inside the CMS database structure and it will be easier to export/backup if needed

**Reply by MikeAlhayek** [reply]
I don't need to render the ID on the create. The Id is generated during the [post request or saving process](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Controllers/LocationController.cs#L92-L127). In fact the entire create process is working as expected. In other words, I am able to create new locations with no problems. 

The problem However, is with the edit screen. To start from the top, the [GET request](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Controllers/LocationController.cs#L129-L148) to the edit screen successfully passes the `Id` as expected. Using the given `Id` I am able to fetch the corresponding model using my `UnitOfWork` then use the `DisplayManager` to create a shape of the [EditLocationViewModel](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Drivers/LocationDisplayDriver.cs#L32-L42). Note on [line 39](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Drivers/LocationDisplayDriver.cs#L38) `viewModel.Id` is set correctly. Thus far in the cycle, there is no problems. 

The next step is to create the `Edit` form in the [Edit.cshtml](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Views/Locations/Edit.cshtml) view. As you can see in the [Edit.cshtml](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Views/Locations/Edit.cshtml#L10-L14) I create the form wrapper. This form set the action to POST the data to the `Location/Edit/{id}` on submitting. To create the body of the form, I am using `DisplayAsync(Model.Content)` to render the `EditLocationViewModel` shape inside the form wrapper. The *problem* is in the `Edit.cshtml` as the `Model.Id` on line [10](https://github.com/CrestApps/CrestApps.Orchard/blob/main/Crestapps.Orchard.PointOfSale/Views/Locations/Edit.cshtml#L10) is always null because `Model.Id` is the id for the shape which is generated outside my process. To fix this problem, I need a way to access the `Id` property located inside the `Model.Content` shape which represents the `EditLocationViewModel`. Do you see where the issue is at now?
